pallets/liquidation/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+
+		/// Account who receives the profit.
+		#[pallet::constant]
+		type ProfitReceiver: Get<Self::AccountId>;
+
+		/// Router weight information.
+		type RouterWeightInfo: AmmTradeWeights<Trade<AssetId>>;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::type_value]
+	pub fn DefaultBorrowingContract() -> EvmAddress {
+		EvmAddress::from_slice(hex_literal::hex!("1b02E051683b5cfaC5929C25E84adb26ECf87B38").as_slice())
+	}
+
+	/// Borrowing market contract address
+	#[pallet::storage]
+	pub type BorrowingContract<T: Config> = StorageValue<_, EvmAddress, ValueQuery, DefaultBorrowingContract>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Money market position has been liquidated
+		Liquidated {
+			liquidator: T::AccountId,
+			evm_address: EvmAddress,
+			collateral_asset: AssetId,
+			debt_asset: AssetId,
+			debt_to_cover: Balance,
+			profit: Balance,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// AssetId to EVM address conversion failed
+		AssetConversionFailed,
+		/// Liquidation call failed
+		LiquidationCallFailed,
+		/// Provided route doesn't match the existing route
+		InvalidRoute,
+		/// Liquidation was not profitable enough to repay flash loan
+		NotProfitable,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T::AccountId: AsRef<[u8; 32]> + IsType<AccountId32>,
+	{
+		/// Liquidates an existing money market position.
+		///
+		/// Performs a flash loan to get funds to pay for the debt.
+		/// Received collateral is swapped and the profit is transferred to `FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: Signed origin.
+		/// - `collateral_asset`: Asset ID used as collateral in the MM position.
+		/// - `debt_asset`: Asset ID used as debt in the MM position.
+		/// - `user`: EVM address of the MM position that we want to liquidate.
+		/// - `debt_to_cover`: Amount of debt we want to liquidate.
+		/// - `route`: The route we trade against. Required for the fee calculation.
+		///
+		/// Emits `Liquidated` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::liquidate()
+			.saturating_add(<T as Config>::RouterWeightInfo::sell_weight(route))
+			.saturating_add(<T as Config>::GasWeightMapping::gas_to_weight(<T as Config>::GasLimit::get(), true))
+		)]
+		pub fn liquidate(
+			origin: OriginFor<T>,
+			collateral_asset: AssetId,
+			debt_asset: AssetId,
+			user: EvmAddress,
+			debt_to_cover: Balance,
+			route: Vec<Trade<AssetId>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let pallet = Self::account_id();
+
+			let debt_original_balance = <T as Config>::Currency::balance(debt_asset, &pallet);
+			let collateral_original_balance = <T as Config>::Currency::balance(collateral_asset, &pallet);
+
+			// mint debt asset
+			<T as Config>::Currency::mint_into(debt_asset, &pallet, debt_to_cover)?;
```

Comment:
what if i specify debt_to_cover to hit the MAX value of u128? 

any issues with that ?

Reply:
should be handled by balances/tokens pallet

Reply:
yes it is handled. but i really meant - does it have any side effects ? 

i cant think of one, but maybe somewhere we use total issuance to calculate something , so this would make a difference

Reply:
We don't change total supply (supply after the execution is the same as before). So I don't see any problem.